src/wallet/wallet.cpp

```diff
@@ -3349,14 +3365,62 @@ void CReserveKey::ReturnKey()
     vchPubKey = CPubKey();
 }
 
+void CWallet::MarkReserveKeysAsUsed(const CKeyID& keyId)
+{
+    LOCK(cs_wallet);
+    CWalletDB walletdb(*dbw);
+    auto it = std::begin(setKeyPool);
+
+    bool foundInternal = false;
+    int64_t foundIndex = -1;
+    for (const int64_t& id : setKeyPool) {
+        CKeyPool keypool;
+        if (!walletdb.ReadPool(id, keypool))
+            throw std::runtime_error(std::string(__func__) + ": read failed");
+
+        if (keypool.vchPubKey.GetID() == keyId) {
```

Comment:
It would also require to keep the `fInternal` flags in memory (for the address book internal/external change flagging). This is independently addresses in #10238.